execution/gethexec/sync_monitor.go

```diff
@@ -0,0 +1,106 @@
+package gethexec
+
+import (
+	"context"
+
+	"github.com/offchainlabs/nitro/execution"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+)
+
+type SyncMonitorConfig struct {
+	SafeBlockWaitForBlockValidator      bool `koanf:"safe-block-wait-for-block-validator"`
+	FinalizedBlockWaitForBlockValidator bool `koanf:"finalized-block-wait-for-block-validator"`
+}
+
+var DefaultSyncMonitorConfig = SyncMonitorConfig{
+	SafeBlockWaitForBlockValidator:      false,
+	FinalizedBlockWaitForBlockValidator: false,
+}
+
+func SyncMonitorConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".safe-block-wait-for-block-validator", DefaultSyncMonitorConfig.SafeBlockWaitForBlockValidator, "wait for block validator to complete before returning safe block number")
+	f.Bool(prefix+".finalized-block-wait-for-block-validator", DefaultSyncMonitorConfig.FinalizedBlockWaitForBlockValidator, "wait for block validator to complete before returning finalized block number")
+}
+
+type SyncMonitor struct {
+	config    *SyncMonitorConfig
+	consensus execution.ConsensusInfo
+	exec      *ExecutionEngine
+}
+
+func NewSyncMonitor(config *SyncMonitorConfig, exec *ExecutionEngine) *SyncMonitor {
+	return &SyncMonitor{
+		config: config,
+		exec:   exec,
+	}
+}
+
+func (s *SyncMonitor) SyncProgressMap() map[string]interface{} {
+	res := s.consensus.SyncProgressMap()
+	consensusSyncTarget := s.consensus.SyncTargetMessageCount()
+
+	built, err := s.exec.HeadMessageNumber()
+	if err != nil {
+		res["headMsgNumberError"] = err
+	}
+
+	if built+1 >= consensusSyncTarget && len(res) == 0 {
+		return res
+	}
+
+	res["builtBlock"] = built
+	res["consensusSyncTarget"] = consensusSyncTarget
```

Comment:
I think it'd still be useful to include all the other info that the consensus side has at this point, even if the consensus side thought it was synced. Is it possible the consensus side thinks it's synced but the execution side doesn't? I think making the SyncProgressMap API slightly more complicated would be helpful, e.g. either relying on the consensus's view of whether or not we're synced instead of determining it above, or have a flag we can pass to SyncProgressMap that forces it to produce the unsynced output even if it thinks it's synced, and we set that flag to true if the execution side thinks it's unsynced. There's a number of versions of this API possible, and I don't have too much of a specific preference besides the fact that it's a bit unfortunate if we only get back sync info from one side but not the other.

Reply:
agreed. done